### PR TITLE
Add ProximityArchive for novelty search

### LINKED ISSUE #468: Support unstructured archives [FEATURE REQUEST]
### Description

I am working on an algorithm which relies on an unstructured archive. This means the archive doesn't have a grid structure which is known ahead of time. Instead, the archive can grow unbounded and uses a "nearest neighbor" approach to add solutions. Unfortunately, the paper which describes this archive doesn't give a reference to where this type of archive originated so I can't provide a source. The paper I am referencing is here: https://dl.acm.org/doi/abs/10.1145/3583131.3590524

The archive uses a distance parameter $\lambda$ to determine if there is an entry in the "$\lambda$-ball" in measure space and adds the solution in the usual way. This is probably most similar to a CVT archive where the number and locations of the centroids are updated every time a solution is added... For a CVT archive recalculating the centroids can be resource intensive. The unstructured archive differs from CVT and Grid Archives in that the bounds and partitioning of metric space don't need to be known in advance, the number of cells is non-deterministic, and there is no need to calculate centroids for every update.

The specific unstructured archive mentioned here uses a Euclidean distance to determine the cells and whether or not objectives should be compared but this could be made more general to use other methods, such as a density-based approach (if the density of solutions in region of measure space is below a threshold then add the solution unconditionally, otherwise replace the lowest objective solution in the region), energy-based approach, etc.

### PR DESCRIPTION
**Pull request by gresavage**
## Description

Implement the ProximityArchive (also known as unstructured archive or novelty archive) from Novelty Search (Lehman, 2011) http://eplex.cs.ucf.edu/papers/lehman_ecj11.pdf

Resolves #468

## Naming

`ProximityArchive` is a specific name that reflects how the archive operations are based on how close solutions are to each other in measure space. We also considered `NearestNeighborArchive`, but that name is a bit long. We find that `NoveltyArchive` and `UnstructuredArchive` are rather imprecise terms.

## TODO

- [x] Add tests -> also added python-box as a dev dependency in setup.py to make tests easier
- [x] Implement class
- [x] Allow passing in objective=None -> objective will then default to 0
- [x] Add tests for scheduler with None objective (also fixed handling of None objectives for `add_mode="single"`)
- [x] Tidy up TODOs in class (i.e., clean up code)

## Status

- [x] I have read the guidelines in
      [CONTRIBUTING.md](https://github.com/icaros-usc/pyribs/blob/master/CONTRIBUTING.md)
- [x] I have formatted my code using `yapf`
- [x] I have tested my code by running `pytest`
- [x] I have linted my code with `pylint`
- [x] I have added a one-line description of my change to the changelog in
      `HISTORY.md`
- [x] This PR is ready to go

### CODE REVIEW

File: ribs/archives/_unstructured_archive.py [hunk 1]
```diff
@@ -0,0 +1,524 @@
...
+        also calculated and concatenated with the distances from before.
+
+        Next, we compute the mean distance of the k-nearest neighbors for each
+        incoming measure. If any of the mean distances are *above* the
+        `sparsity_threshold` then the archive is resized to accomodate them.
+        They are each assigned new indices in the archive.
+
+        Args:
+            measures (array-like): (batch_size, :attr:`measure_dim`) array of
+                coordinates in measure space.
```
Comment (btjanaka) [inline ribs/archives/_unstructured_archive.py:125]: What does resize do? I'd prefer not to change the API of index_of if possible.

File: ribs/archives/_unstructured_archive.py [hunk 2]
```diff
@@ -0,0 +1,524 @@
...
+    @property
+    def sparsity_threshold(self) -> float:
+        """:attr:`dtype` : The degree of sparseness in metric space required for
+        a solution to be added unconditionally."""
+        return self._sparsity_threshold
+
+    def index_of(self, measures, resize: bool = False) -> np.ndarray:
+        """Returns archive indices for the given batch of measures.
+
+        First, if the archive is empty it is resized to fit the new data.
```
Comment (btjanaka) [inline ribs/archives/_unstructured_archive.py:110]: In general, batch insertion behavior for novelty search has not really been studied before; prior works assume you are only adding one solution at a time. For example, the Archive class in QDax just defines an iteration function and uses `jax.lax.scan` to apply it over the entire batch of new solutions https://github.com/adaptive-intelligent-robotics/QDax/blob/eb2cf317c7a5b6ffe385f0606a99e7c509b771b3/qdax/core/containers/archive.py#L134 Thus, we have some leeway in how we choose to handle things, but it should all be an extension of the original novelty search mechanisms, namely, if we insert one solution at a time with this method, we should get the same behavior as the original novelty search.

That being said, for the initial insertion, I don't think we should just insert all the solutions at once. That would mean we have solutions that fail to satisfy the novelty threshold and are too close together.

For subsequent insertions, we can find the k-nearest neighbors among both the archive and the incoming measures, and only pick solutions that satisfy the novelty threshold. I think this is what you are saying with computing distances to archive solutions and other incoming solutions and concatenating the distances. However, it can be put more succinctly by taking the whole set of `archive + incoming measures` together. When we look at it this way, the initial insertion is actually just the same as subsequent insertions, as the initial insertion will just have an incoming archive. That should help simplify this method.

Finally, I think you can make things a lot cleaner and faster by using scipy's cKDTree (or just KDTree nowadays), specifically the query method: https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.cKDTree.query.html#scipy.spatial.cKDTree.query

**Review by btjanaka — CHANGES_REQUESTED** [review]
Hi @gresavage, thank you for your PR! I appreciate that it is well-documented and tested and that you clearly put a lot of time into it. I have left comments requesting a couple of changes.

Overall, I think the UnstructuredArchive is great. The UnstructuredGridArchive seems rather specialized, so I do not think we should add it to pyribs. Could you remove from this PR the UnstructuredGridArchive and the roll() method in ArrayStore?

I know these things can take quite a bit of time; if you need to work on other things, just let me know and I'm happy to help make the changes I mentioned.

File: ribs/archives/_array_store.py [hunk 3]
```diff
@@ -538,6 +538,29 @@ def resize(self, capacity):
             self._fields[name] = np.empty(new_shape, cur_arr.dtype)
             self._fields[name][:cur_capacity] = cur_arr
 
+    def roll(self, shift):
```
[the remark this answers is not in the document]
Comment (gresavage) [reply ribs/archives/_array_store.py:541]: I suppose we could eliminate the need for this entirely by changing how the indices are determined. If we do away with using the bounds & interval size to determine the index then the archive would be able to grow by simply appending indices... the trade-off is that the operation for calculating the indices would increase in computational complexity.

Another possibility is to implement both, but make a kwarg for `GridUnstructuredArchive.__init__` (or other classes which might use `roll`) to determine whether to use the bounds/interval size + roll or whether to use the alternate lookup-based method.

File: ribs/archives/_unstructured_archive.py [hunk 4]
```diff
@@ -0,0 +1,524 @@
...
+
+    This archive is described in `Lehman 2011
+    <https://www.cs.swarthmore.edu/~meeden/DevelopmentalRobotics/lehman_gecco11.pdf>`_.  # pylint: disable=C0301
+    If a solution is in a sparse area of metric space it is added
+    unconditionally. When a solution is in an overdense region it is added to
+    the archive only if its objective improves upon the nearest existing
+    solution. The archive uses the mean distance of the k-nearest neighbors to
+    determine the sparsity of metric space
+
+    .. note:: The archive is initialized as empty
```
[the remark this answers is not in the document]
Comment (gresavage) [reply ribs/archives/_unstructured_archive.py:27]: No, I think I put that there to allude to the interactions with other methods which check for `archive.empty` or the fact that there might not be `k`-neighbors... I can remove this if you like

File: tests/archives/conftest.py [hunk 5]
```diff
@@ -140,17 +154,32 @@ def get_archive_data(name, dtype=np.float64):
...
+                                                 sparsity_threshold=1.0,
+                                                 dtype=dtype)
+        grid_indices = (0,)
+        int_index = 0
+
+    archive_with_elite.add_single(solution, objective, measures)  # pylint: disable=E
 
     return ArchiveFixtureData(
-        archive,
+        archive,  # pylint: disable=E
```
[the remark this answers is not in the document]
Comment (gresavage) [reply tests/archives/conftest.py:176]: This must be an accident - I always mean to put the full code. In this case it was [E0606](https://pylint.readthedocs.io/en/stable/user_guide/messages/error/possibly-used-before-assignment.html)

That was being flagged by pre-commit/pylint because `archive` is only defined inside conditional statements and there is no `else` clause defining it. I ignored it because this was the same structure/behavior you had before, and if `archive` is not defined the test would fail anyway.

File: ribs/archives/_unstructured_archive.py [hunk 6]
```diff
@@ -0,0 +1,524 @@
...
+
+    Args:
+        solution_dim (int): Dimension of the solution space.
+        measure_dim (int): The dimension of the measure space.
+        k_neighbors (int): The number of nearest neighbors to use for
+            determining sparseness.
+        sparsity_threshold (float): The level of sparsity required to add a
```
[the remark this answers is not in the document]
Comment (gresavage) [reply ribs/archives/_unstructured_archive.py:34]: I used sparsity to tie it more closely to the concepts of distance and density but I am happy to replace it

File: ribs/archives/_unstructured_archive.py (hunk 1 above)
Comment (gresavage) [reply ribs/archives/_unstructured_archive.py:125]: `resize` indicates whether or not the archive is allowed to grow to support any solutions which match the novelty criterion, or whether or not to simply return the index of the nearest neighbor.

I did this partly so that the archive didn't grow in memory unless we specifically intended to add solutions. I too didn't want to change the API in this way so I can remove the check.

File: ribs/archives/_unstructured_archive.py [hunk 7]
```diff
@@ -0,0 +1,524 @@
...
+                self._store.resize(curr_occ + additions)
+                self._cells = self._store.capacity
+
+            # assign new entries to unoccupied indices
+            free_indices = list(np.argwhere(~self._store.occupied))
+            for unq in unique_indices:
+                cell_map[unq] = free_indices.pop(0).item()
+        return np.array([cell_map.get(s, s) for s in indices])
+
+    def index_of_single(self, measures, resize: bool = False) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (gresavage) [reply ribs/archives/_unstructured_archive.py:227]: yes - when I do away with the `resize` argument `add`, `add_single`, and `index_of_single` will inherit from `ArchiveBase`

File: ribs/archives/_unstructured_archive.py [hunk 8]
```diff
@@ -0,0 +1,524 @@
...
+
+        for name, arr in add_info.items():
+            add_info[name] = arr[0]
+
+        if add_info["status"]:
+            self._stats_update(objective_sum, best_index)
+
+        return add_info
+
+    @cached_property
```
[the remark this answers is not in the document]
Comment (gresavage) [reply ribs/archives/_unstructured_archive.py:457]: I used `cached_property` so that `np.max` and `np.min` didn't have to be called if the archive didn't change since they are called on the entire archive. Then the `add` method removes these properties so that the values are re-calculated any time the properties are accessed again after `add` is called.

If you're fine with keeping `cached_property` then I could make this even a little more efficient by only clearing the cache if there are `2`s in the status batch.

File: ribs/archives/_unstructured_archive.py [hunk 9]
```diff
@@ -0,0 +1,524 @@
...
+
+
+class UnstructuredArchive(ArchiveBase):
+    """An archive that adds new solutions based on their novelty.
+
+    This archive is described in `Lehman 2011
+    <https://www.cs.swarthmore.edu/~meeden/DevelopmentalRobotics/lehman_gecco11.pdf>`_.  # pylint: disable=C0301
+    If a solution is in a sparse area of metric space it is added
+    unconditionally. When a solution is in an overdense region it is added to
+    the archive only if its objective improves upon the nearest existing
```
[the remark this answers is not in the document]
Comment (gresavage) [reply ribs/archives/_unstructured_archive.py:23]: Kind of - the paper by Lehman I referenced includes an objective but it is actually not what I described here so I will change the code to more closely match the original work. I think this means a new transform or two since rather than just comparing objective values, they use a method more akin to non-dominating sort of three objectives:

1. novelty - sparsity in measure space
2. genotypic diversity - uniqueness of network topology
3. fitness - the number of nearby solutions with objective scores less than the solution

A solution is added iff it is better than solutions in its neighborhood on any objective and simultaneously no worse than its neighborhood on all other objectives. So a solution which improves novelty, but "dominates" fewer neighbors in fitness than each of those neighbors would not be added.

At this point, since all networks have essentially the same genotype we are only comparing objectives (1) and (3)

[the remark this answers is not in the document]
Comment (gresavage) [reply ribs/archives/_unstructured_archive.py:23]: Perhaps the best solution instead of doing the kNN search in `index_of` as I initially did, we *only* do it in a transform. The `index_of` would then just return the nearest neighbor already in the archive. It would still be a little expensive as we would be calculating distances twice (once in `index_of` and again in the transform) but cheaper than doing kNN twice.

[the remark this answers is not in the document]
Comment (btjanaka) [reply ribs/archives/_unstructured_archive.py:23]: I see the confusion. Joel and Ken released two papers that year. The first was Novelty Search. The second was Novelty Search with Local Competition, which you linked above. Novelty Search is the one we're looking for: http://eplex.cs.ucf.edu/papers/lehman_ecj11.pdf

[the remark this answers is not in the document]
Comment (btjanaka) [reply ribs/archives/_unstructured_archive.py:23]: Regarding kNN, I think what you have currently is almost the correct logic, and you should not need to modify transforms. Basically, if the solution is novel enough, then just assign it a new index in the archive, and the current add method should work just fine because it will never have to deal with solutions that are in the same index.

If you don't mind, I'd like to take a pass on this over the next few days; I think I have a good idea of how to do it.

**Comment by btjanaka** [comment]
Hi @gresavage, just wanted to give a quick update on things. I'm going through and modifying the UnstructuredArchive -- I've changed it a decent amount to align more with the rest of the library and how we implement things. I removed GridUnstrucutredArchive and ArrayStore.roll as per my earlier comments, but do consider taking a look at my comment here https://github.com/icaros-usc/pyribs/pull/472#discussion_r1649428413 as I think it reveals an unexpected behavior. I'll keep working on this bit by bit, and it should be ready soon. Thanks again!

File: ribs/archives/_array_store.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (gresavage) [reply ribs/archives/_array_store.py]: I'm happy to remove it - I included it solely for the sake of trying to implement unstructured variants of existing archive classes... GridArchive was a "low hanging" fruit in that sense. I too had not seen any kind of grid unstructured archive.

File: ribs/archives/_unstructured_archive.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (gresavage) [reply ribs/archives/_unstructured_archive.py]: Novelty Search alone is certainly much simpler to implement than what I was stepping into. However, I think it would be nice to include both behaviors. The authors seemed to indicate local competition exceeded both novelty alone, and global competition.

My only concern with slotting a novel solution into a new index is how quickly the memory footprint might grow. Maybe you have an idea of how to do this efficiently.

I am curious to see what you come up with.

I am basically done implementing the NS with local competition. The Stanley and Lehman treat the problem as a MOEA where the objectives are fitness and genotypic variation. They take the NSGA-II approach but replace the "crowding distance" with "novelty" as the authors suggest. To that end I made the archive class able to specify additional objectives to include as part of the MOEA problem with the default behavior being single objective: maximizing fitness.

Also, to that end, I implemented some non-dominating sort algorithms... however I think covering NDS algorithms would expand the scope of the PyRibs project considerably. If including NS + local competition is desired, then it might be best to add a dependency on another project where NDS algorithms are in-scope. These are some projects I know of which cover NDS algorithms, PyMOO is the best that I'm aware of:

- [PyGAD](https://pygad.readthedocs.io/en/latest/) | [code](https://github.com/ahmedfgad/GeneticAlgorithmPython/blob/master/pygad/utils/nsga2.py)
  - this one only implements NDS in python, and only has a single implementation of NDS
- [PyMOO](https://pymoo.org/) | [code](https://github.com/anyoptimization/pymoo/tree/main/pymoo/util/nds)
  - this library has the best support for NDS since it has quite a few algorithms implemented, as well as Cython variants which speed up the algorithm considerably over pure python
- [Sferes2](https://sferes2.github.io/sferes2/) | [code](https://github.com/sferes2/sferes2/blob/master/sferes/ea/dom_sort.hpp)
  - this is a C++ library so including it would be a major hurdle, it only implements a single NDS algorithm, and hasn't been updated in nearly a decade... I included it simply because quite a few research papers and other projects use it or reference it.
- [Pygmo](https://esa.github.io/pygmo2/index.html) | [code](https://github.com/esa/pygmo2/blob/master/pygmo/core.cpp#L405)
  - this is a mixed C++/Python library which is a Python port of [Pagmo2](https://esa.github.io/pagmo2/). Its major drawback is that not only does it require the Pagmo2 C++ library be installed, but just a single NDS algorithm is implemented

**Comment by gresavage** [comment]
Hi @btjanaka I'm eager to see what you've done.

I haven't pushed any of my changes w.r.t NS + local competition yet as they are still in flux and I'd like to know your thoughts on handling non-dominated sorting... I also want to fold my changes into what you've got so the two strategies are homogeneous if you decide to include NS + local competition.

I was also thinking more about the objective threshold and learning rate. Are we sure we want to do away with this behavior entirely if we implement NS + local competition? I removed the keyword arguments in my last commit but I think it could be useful to have CMA-MAE behavior with the local competition variant.

**Comment by btjanaka** [comment]
> Hi @btjanaka I'm eager to see what you've done.
> 
> I haven't pushed any of my changes w.r.t NS + local competition yet as they are still in flux and I'd like to know your thoughts on handling non-dominated sorting... I also want to fold my changes into what you've got so the two strategies are homogeneous if you decide to include NS + local competition.
> 
> I was also thinking more about the objective threshold and learning rate. Are we sure we want to do away with this behavior entirely if we implement NS + local competition? I removed the keyword arguments in my last commit but I think it could be useful to have CMA-MAE behavior with the local competition variant.

@gresavage I'll reply to both your comments here.

To implement NSLC in pyribs, I think we would create an emitter that contains the MOEA, most likely using one of the libraries you mentioned (indeed, it would likely be out of scope to create NDS algorithms in pyribs unless there is some custom feature we want). The archive would then output the novelty and local competition via the add_info returned from the add() method, and the emitter would use the NDS to rank based on the novelty and local competition. Alternatively, if the emitter needs to also include the population in novelty and local competition computations, then it can compute those metrics inside of the emitter rather than in the archive.

Regarding the archive, I took another look at the NSLC paper, and I think one point that may be confusing us (at least me) is whether solutions are replaced based on their local competition. My understanding is that NSLC just uses a regular archive identical to the one from novelty search. The local competition seems to be a metric that is computed using the solutions in the archive, but it does not seem the metric plays any role in updating the archive. If you saw a section or piece of code saying otherwise, could you send it along? I've also asked some friends in the QD community for their thoughts on this; I'm genuinely not sure how this part works.

If I am right, then the current archive would be roughly what we need for both NS and NSLC, and the job of implementing NSLC would likely involve making a new emitter as I described above.

File: ribs/archives/_unstructured_archive.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (btjanaka) [reply ribs/archives/_unstructured_archive.py]: (Resolving; responded in general thread)

**Comment by gresavage** [comment]
> > Hi @btjanaka I'm eager to see what you've done.
> > I haven't pushed any of my changes w.r.t NS + local competition yet as they are still in flux and I'd like to know your thoughts on handling non-dominated sorting... I also want to fold my changes into what you've got so the two strategies are homogeneous if you decide to include NS + local competition.
> > I was also thinking more about the objective threshold and learning rate. Are we sure we want to do away with this behavior entirely if we implement NS + local competition? I removed the keyword arguments in my last commit but I think it could be useful to have CMA-MAE behavior with the local competition variant.
> 
> @gresavage I'll reply to both your comments here.
> 
> To implement NSLC in pyribs, I think we would create an emitter that contains the MOEA, most likely using one of the libraries you mentioned (indeed, it would likely be out of scope to create NDS algorithms in pyribs unless there is some custom feature we want). The archive would then output the novelty and local competition via the add_info returned from the add() method, and the emitter would use the NDS to rank based on the novelty and local competition. Alternatively, if the emitter needs to also include the population in novelty and local competition computations, then it can compute those metrics inside of the emitter rather than in the archive.
> 
> Regarding the archive, I took another look at the NSLC paper, and I think one point that may be confusing us (at least me) is whether solutions are replaced based on their local competition. My understanding is that NSLC just uses a regular archive identical to the one from novelty search. The local competition seems to be a metric that is computed using the solutions in the archive, but it does not seem the metric plays any role in updating the archive. If you saw a section or piece of code saying otherwise, could you send it along? I've also asked some friends in the QD community for their thoughts on this; I'm genuinely not sure how this part works.
> 
> If I am right, then the current archive would be roughly what we need for both NS and NSLC, and the job of implementing NSLC would likely involve making a new emitter as I described above.


#### RE: Emitters

I think an emitter could work for NSLC so long as the ranking doesn't play a roll in what gets added to the archive. The only real hurdle I see with that route is that NSLC doesn't bother with how candidates are generated, so an NSLC emitter could essentially wrap any of the other emitters and just leave the `ask` method unchanged. In my eyes this means making an NSLC variant for each of the other emitters which is only a hurdle inasmuch as it is tedious :laughing: . The `tell` method would be where the magic happens.


#### RE: Replacement & Use of NDS Rank

I looked over the papers as well and I think you're right that solutions are not replaced - which greatly simplifies the issue. However, my understanding of NSLC is that the rank (i.e. which Pareto front the solution occupies) replaces the novelty metric in NS. Which then leads me to think that if NS modifies the objective by replacing it with novelty, and NSLC modifies NS by replacing novelty with rank, then the NDS rank plays a roll in updating the archive. The NSLC paper never mentions explicitly how solutions are added to the archive - whether it is based on novelty alone or their domination rank. If they are added based solely on novelty, then the archive will never add new solutions once a region of space drops below the novelty threshold even if a new solution out-performs existing solutions.

This would seem to lead us back to the original issue NSLC is trying to address where only keeping novel solutions doesn't guarantee the solutions functional. Even if an evolutionary algorithm uses the rank and modifies the genome, the emitter will always be pulling from the first elites which map to that region of space and so the current population would essentially only be a single evolutionary epoch more evolved than the archive. Hence, when determining what to add to the archive for NSLC I think the NDS ranking needs to be determined first, and if the solution outranks its nearest neighbors it should be assigned a new index.

If that's the case, then I don't think an emitter would work for NSLC since ranking would need to happen before the archive's `add` method is called.

**Comment by btjanaka** [comment]
I see what you are saying about using the NSLC ranking. I think a good path would be to implement something in `ribs.emitters.rankers`. Essentially, the ranker would integrate an NDS, and the ranker could be used in emitters like EvolutionStrategyEmitter. Meanwhile, the archive could return `local_competition` in `add_info`. The emitter would take in the novelty and local_competition from the add_info and pass these pieces of info to the ranker.

For now, I've updated the code to just be a regular Novelty Search archive. I added a TODO to add local competition on issue #474. Would you mind taking a look at the code? Thanks!

**Comment by gresavage** [comment]
> @gresavage I see what you are saying about using the NSLC ranking. I think a good path would be to implement something in `ribs.emitters.rankers`. Essentially, the ranker would integrate an NDS, and the ranker could be used in emitters like EvolutionStrategyEmitter. Meanwhile, the archive could return `local_competition` in `add_info`. The emitter would take in the novelty and local_competition from the add_info and pass these pieces of info to the ranker.
> 
> For now, I've updated the code to just be a regular Novelty Search archive. I added a TODO to add local competition on issue #474. Would you mind taking a look at the code? Thanks!

@btjanaka AHHHH, rankers, of course! Clever :)
image

I will take a look.

File: ribs/archives/_nearest_neighbor_archive.py [hunk 10]
```diff
@@ -0,0 +1,489 @@
...
+        reference_measures = self.data("measures")
+        if self._compare_to_batch:
+            reference_measures = np.concatenate(
+                (reference_measures, data["measures"]), axis=0)
+
+        if self.compare_to_batch:
+            # In this case, solutions from the batch will be included in the
+            # nearest neighbors. Thus, the first neighbor of each solution will
+            # be the solution itself, so we need to compute an extra neighbor
+            # and exclude the first neighbor.
```
Comment (gresavage) [inline ribs/archives/_nearest_neighbor_archive.py:297]: I'm not sure the statement "the first neighbor of each solution will be the solution itself".

A counterexample:
In the unlikely event that two solutions share the exact same point measure space and unless `cKDTree` behaves differently than other sort methods, the lower-indexed one will appear first. This means the "zeroth" neighbor is not always the solution itself.

In an even more unlikely scenario where `k+1` other solutions occupy the same point in measure space *and* all have lower indices than the solution itself, then the solution won't even appear in the k-nearest neighbors.

---

I went about handling this by filtering the return of `cKDTree.query` with something like `knn_indices != (len(self) + np.arange(batch_size))[:, None]`

File: ribs/archives/_nearest_neighbor_archive.py [hunk 11]
```diff
@@ -0,0 +1,489 @@
...
+        n_eligible = np.sum(eligible)
+        new_size = len(self) + n_eligible
+
+        if new_size > self.capacity:
+            # Resize the store by doubling its capacity. We may need to double
+            # the capacity multiple times. The log2 below indicates how many
+            # times we would need to double the capacity. We obtain the final
+            # multiplier by raising to a power of 2.
+            multiplier = 2**int(np.ceil(np.log2(new_size / self.capacity)))
+            self._store.resize(multiplier * self.capacity)
```
Comment (gresavage) [inline ribs/archives/_nearest_neighbor_archive.py:332]: it is possible that the user sets the initial capacity to 0, therefore the archive would never grow using this strategy.

I think it's simplest to just throw an `initial_capacity = max(1, initial_capacity)` in `__init__`

**Comment by gresavage** [comment]
@btjanaka Do you think we need to implement visualization for the archive?

**Comment by btjanaka** [comment]
> @btjanaka Do you think we need to implement visualization for the archive?

Yes, I added a TODO on #474

**Comment by gresavage** [comment]
> > @btjanaka Do you think we need to implement visualization for the archive?
> 
> Yes, I added a TODO on #474

Apologies! I literally just opened that up :grimacing:

File: ribs/archives/_nearest_neighbor_archive.py (hunk 10 above)
Comment (btjanaka) [reply ribs/archives/_nearest_neighbor_archive.py]: If two solutions are at the same point, then one of them should indeed be marked as the first neighbor and will be excluded, and the other should be marked as a neighbor that should be included.

Anyway, I think this `compare_to_batch` behavior is not a good idea because it essentially forces the batch to be spread out. I will remove it.

File: ribs/archives/_nearest_neighbor_archive.py (hunk 11 above)
Comment (btjanaka) [reply ribs/archives/_nearest_neighbor_archive.py]: Good point! I now throw an error if initial_capacity is less than 1.

**Review by btjanaka — APPROVED** [review]
